schedule-update/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+		pub fn schedule_dispatch(origin, call: CallOf<T>, when: DelayedDispatchTime<T::BlockNumber>) {
+			let who: Option<T::AccountId>;
+			if ensure_signed(origin.clone()).is_ok() {
+				who = ensure_signed(origin.clone()).ok();
+			} else if ensure_root(origin.clone()).is_ok() {
+				who = None;
+			} else {
+				return Err(Error::<T>::BadOrigin.into());
+			}
+
+			let id = Self::_get_next_id()?;
+
+			let block_number = match when {
+				DelayedDispatchTime::At(block_number) => {
+					block_number
+				},
+				DelayedDispatchTime::After(block_count) => {
+					<frame_system::Module<T>>::block_number() + block_count
+				},
+			};
+
+			match call.get_dispatch_info().class {
+				DispatchClass::Normal => {
+					<DelayedNormalDispatches<T>>::insert(block_number, id, (who, call, id));
+				},
+				DispatchClass::Operational => {
+					<DelayedOperationalDispatches<T>>::insert(block_number, id, (who, call, id));
+				},
+			}
+			Self::deposit_event(RawEvent::ScheduleDispatch(block_number, id));
+		}
+
+		/// Cancel schedule_update
+		pub fn cancel_deplayed_dispatch(origin, at: T::BlockNumber, id: DispatchId) {
+			let is_root = ensure_root(origin.clone()).is_ok();
+
+			if <DelayedNormalDispatches<T>>::contains_key(at, id) {
+				if !is_root {
+					let w = ensure_signed(origin)?;
+					let (who, _, _) = <DelayedNormalDispatches<T>>::get(at, id);
+					if Some(w) != who {
+						return Err(Error::<T>::NoPermission.into());
+					}
+				}
+				<DelayedNormalDispatches<T>>::remove(at, id);
+			} else if <DelayedOperationalDispatches<T>>::contains_key(at, id) {
+				if !is_root {
+					let w = ensure_signed(origin)?;
+					let (who, _, _) = <DelayedOperationalDispatches<T>>::get(at, id);
+					if Some(w) != who {
+						return Err(Error::<T>::NoPermission.into());
+					}
+				}
+				<DelayedOperationalDispatches<T>>::remove(at, id);
+			} else {
+				return Err(Error::<T>::DispatchNotExisted.into());
+			}
+			Self::deposit_event(RawEvent::CancelDeplayedDispatch(id));
+		}
+
+		fn on_initialize(now: T::BlockNumber) {
+			let mut weight: Weight = 0;
+			let total_weight = T::MaxScheduleDispatchWeight::get();
+
+			// Operational calls are dispatched first and then normal calls
+			// TODO: dispatches should be sorted
+			let operational_dispatches = <DelayedOperationalDispatches<T>>::iter_prefix(now);
+			operational_dispatches.for_each(|(who, call, id)| {
+				weight += call.get_dispatch_info().weight;
+				if weight > total_weight {
+					return;
+				}
+
+				let origin: T::Origin;
+				if let Some(w) = who {
+					origin = frame_system::RawOrigin::Signed(w).into();
+				} else {
+					origin = frame_system::RawOrigin::Root.into();
+				}
+
+				let result = call.dispatch(origin.clone());
+				if let Err(e) = result {
+					 Self::deposit_event(RawEvent::ScheduleDispatchFail(id, e));
+				} else {
+					 Self::deposit_event(RawEvent::ScheduleDispatchSuccess(now, id));
+				}
+				<DelayedOperationalDispatches<T>>::remove(now, id);
```

Comment:
use `remove_prefix` to remove all elements at once

Reply:
`remove_prefix ` can't move extra ones to next block